src/Build.UnitTests/Graph/ParallelWorkSet_Tests.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.Build.UnitTests;
+using Shouldly;
+using Xunit;
+
+namespace Microsoft.Build.Experimental.Graph.UnitTests
+{
+    public class ParallelWorkSet_Tests
+    {
+        private class ParallelWorkSetTestCase
+        {
+            internal int DegreeOfParallelism { get; set; }
+            internal List<WorkItem> WorkItemsToAdd { get; set; } = new List<WorkItem>();
+
+            internal Dictionary<string, string> ExpectedCompletedWork =
+                new Dictionary<string, string>(StringComparer.Ordinal);
+            internal bool ShouldExpectException { get; set; }
+        }
+
+        private struct WorkItem
+        {
+            internal string Key { get; set; }
+
+            internal Func<string> WorkFunc { get; set; }
+        }
+
+        private ParallelWorkSet<string, string> _workSet;
+
+        [Fact]
+        public void GivenExceptionsOutsideWorker_CompletesAndThrowsException()
```

Comment:
What does `OutsideWorker` mean here? "Executed inline, rather than on a thread?"

Reply:
Clarified this with ```WorkerThread``` and ```CompletionThread```